[stdlib] `String.join()` and `StringLiteral.join()` ~x1300 speed-up

msaelices opened the pull request:
Great optimizations in the `String.join()` logic, with impressive **~x1300 speed-up** in very large joined lists.

## Strategy

Calculate the total size of the joined string beforehand to reserve the buffer with the right capacity to prevent many syscalls for allocating memory.

## Benchmark

Before:
``` 
❯ mojo bench_join.mojo 
Len:  588889 Time:  1.792499018 seconds
```

After:
``` 
❯ mojo bench_join.mojo 
Len:  588889 Time:  0.001283047 seconds
```

Code used for the benchmark:

```mojo
from time import now

fn main():
    l = List[String]()
    for i in range(100_000):
        l.append(str(i))
    start = now()
    s = String(",").join(l)
    end = now()
    print('Len: ', len(s), 'Time: ', (end - start) / 1_000_000_000, 'seconds')
```

Out of curiosity, this benchmark is now roughly as fast as Python one (implemented in C)

jjvraw committed:
[External] [stdlib] Add negative test for `String.split()`(#47145)

[External] [stdlib] Add negative test for `String.split()`

Add a test for `String.split()` with an empty separator.

Co-authored-by: Joshua James Venter <venter.joshua@gmail.com>
Closes modularml/mojo#3461
MODULAR_ORIG_COMMIT_REV_ID: 6529ff9a90d9bc9eabf6b870c51261ab2e33a0f5

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

KenJones-Modular committed:
Updated v24.5 release date.
Incorporated feedback.
Added an introduction to the Highlights section.
Fixed various typos and formatting issues.

MODULAR_ORIG_COMMIT_REV_ID: bd597d5eedec9a9398a9d34c5838fe8e9074d0b2

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

judyheflin committed:
[mojo-stdlib][docs] Updated mojo docs to import from generalized modules

Updated docs in modules to suggest `from module import foo` instead of
`from package.module import foo`.

MODULAR_ORIG_COMMIT_REV_ID: f4cb0aab668c9b492c2d1ccffd24548a6bea3e21

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

JoeLoser committed:
[stdlib] Make `InlineArray` work on `CollectionElement`

As we're unwinding some of the explicit copyability work to get out of the
halfway state, make `InlineArray[T, Size]` work in terms of `CollectionElement`
and not `CollectionElementNew`.  This will allow things to compose a bit nicer
in the short-term instead of pushing `CollectionElementNew` around even wider
right now.  For example, this would enable `InlineFixedVector` to work in terms
of `CollectionElement` now since internally it would hold an `InlineArray` data
member.

MODULAR_ORIG_COMMIT_REV_ID: cf9877f7c26b57acbcc5a0e66c782a23911a14af

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

JoeLoser committed:
[stdlib] Ensure `not` in `PATH` for Linux builds

Recently, the `not` tool was upgraded to be required and was only installed on
our macOS CI.  Set the alternative for `not` in the Linux build scripts too to
make the Ubuntu builds run the tests that require the `not` tool.

MODULAR_ORIG_COMMIT_REV_ID: bd863f96efcd4fc203736581f19151555b814719

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

gabrieldemarmiesse committed:
[External] [stdlib] Add new method `SIMD._dynamic_shuffle()` (#45892)

[External] [stdlib] Add new method `SIMD._dynamic_shuffle()`

This method is incredibly useful to implement fast UTF-8 validation.
It's used a lot in the paper [Validating UTF-8 In Less Than One
Instruction Per Byte](https://arxiv.org/abs/2010.03090). It notably
allowed me to implement a fast utf8 validation 6 times faster than the
one currently in the stdlib. Since it's a vectorized dynamic
vector-lookup, it can be useful in many other scenarios.

Note that I also tried to implement it with calling the llvm intrinsic
but I would get those kind of errors:

```mojo
    return sys.llvm_intrinsic[
        "llvm.shufflevector",
        SIMD[DType.uint8, 16],
        has_side_effect=False,
    ](lookup_table, indices)
```

`could not find LLVM intrinsic: "llvm.shufflevector"` or `error: could
not find LLVM intrinsic: "shufflevector"` if I remove the llvm prefix.
If that's possible to use it, guidance would be appreciated.

The current implementation works, even if it's not the fastest for some
vector sizes and arch. So I believe if we can't find anything better
quickly, we can merge it and iterate on it later.

Co-authored-by: Gabriel de Marmiesse <gabrieldemarmiesse@gmail.com>
Closes modularml/mojo#3397
MODULAR_ORIG_COMMIT_REV_ID: e2b3095316825a6994ad1c4e023727fa53cbebb6

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

KenJones-Modular committed:
This adds a mention to the Mojo changelog that there is a known issue
when running a program containing input() directly with the mojo CLI.

MODULAR_ORIG_COMMIT_REV_ID: 2f68a7630443ae0509439f481da529a08b4d3b36

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Speed-up String.join() by relying on Formatter

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] New FormattableCollectionElement

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] New String.join() optimization for collection elements

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

modularbot committed:
[stdlib] Bump compiler version to 2024.9.1405

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Optimize String.join() by precalculating alloc size

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Add comment explaining the optimization

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Remove non-needed code

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Microoptimization by calculating the joined bytes beforehand

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Fix the base case of no elements to join

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
[stdlib] More comprehensive comment

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
[stdlib] Change .join() conformance from Stringable to Formattable
This is slightly slower than the Stringable one

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
[stdlib] Adapt the join() logic to base on the as_string_slice() method.
This is way faster but we have a conflict between method conformances

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Better naming conventions

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Remove unused import

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
[stdlib] Micro optimization by not copying the data on variable declaration

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Add a fast parameter to String.join() as a trick
to prevent ambiguity

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
[stdlib] Adapt the as_bytes_slice() code after new *_span() name

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Hack to workaround an "ambiguos call" compiler issue
So, now it will use the fast join if used with strings with no explicit
call.

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

martinvuyk reviewed (commented):
Hi @msaelices great catch this definitely needed optimization :D

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
[stdlib] Use byte_length() to ensure we count the length the same way.
The unicode support may change the len() logic so we need to be careful.
Thanks to @martinvuyk for the suggestion

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] New String.join() speed-up by using memcpy() instead of .extend()
Thanks to @martinvuyk for the suggestion

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Link to the exact Mojo compiler issue we would need to solve

Co-authored-by: martinvuyk <110240700+martinvuyk@users.noreply.github.com>
Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Optimize StringLiteral.join() by reusing String.join() one

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
Save one allocation when we are joining string literals
Thanks to @martinvuyk for the suggestion

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

msaelices committed:
Merge branch 'nightly' into string-join-optimizations

jackos committed:
Updates for changes

jackos committed:
Reword TODO and move compile time check into else block

jackos reviewed (approved):
This is fantastic thanks @msaelices, I learnt from this how to run a different implementation when the concrete type meets a more specific trait conformance.

I reworded the TODO slightly, at the moment this is expected where if you have two overloads on trait types, and the concrete type conforms to both of those traits, it's considered ambiguous as to which function should run. But this is a great workaround until we have something like #3403, I'll be using it myself!

One other thing is I added an else to the `@parameter if` block, so that the IR that won't run is removed at compile time if none of the conditions are True.

Thanks very much for the great PR.

jackos committed:
Fix check_licences path
